models/pull.go

```diff
@@ -1472,3 +1472,46 @@ func TestPullRequests() {
 func InitTestPullRequests() {
 	go TestPullRequests()
 }
+
+// CloseActivePullRequests closes the issue of all the active pull requests associated with a branch
+func CloseActivePullRequests(doer *User, repo *Repository, branchName string) {
+	baseprs, err := GetUnmergedPullRequestsByBaseInfo(repo.ID, branchName)
+	if err != nil {
+		log.Error(4, "Find pull requests [base_repo_id: %d, base_branch: %s]: %v", repo.ID, branchName, err)
+		return
+	}
+	headprs, err := GetUnmergedPullRequestsByHeadInfo(repo.ID, branchName)
+	if err != nil {
+		log.Error(4, "Find pull requests [head_repo_id: %d, head_branch: %s]: %v", repo.ID, branchName, err)
+		return
+	}
+	prs := append(baseprs, headprs...)
+	for statusChanged := true; statusChanged; {
```

Comment:
This is not what I meant by sort the issues.

You have a list of pull-requests, `prs`. Consider two pull-requests `p` and `q`, say `p<q` if `q` has a dependency on `p`. 

Assuming there are no cyclic dependencies, this `<` forms a partial order on the set of PRs in `prs`. You can then sort the list `prs` by this partial order. (There will be multiple valid sortations of the list but who cares.)

You should not have to ask the database potentially n^2 times whether an issue still has dependencies.

Further, this whole section should be done in a transaction - you're at risk of data changing under your feet here.

Reply:
Yeah I had understood what you meant in your comment but after looking at our [code](https://github.com/go-gitea/gitea/blob/b8a81cb1b887799d169fd916a4c4e1a075aca2c3/models/issue_dependency.go#L49) for issue dependency I found that it is indeed possible to create a cycle involving three or more issues like A->B->C->A, in this case sorting can't be done without first detecting and removing the issues involved in the cycle, right? If you are concerned about the database query I can remove that by the replacing the code with some less straightforward code, but shouldn't that database query be cheap since we are performing only a count operation and not actually loading the data?
I think we should fix the issue dependency code anyway to avoid having circular dependency, right?
>Further, this whole section should be done in a transaction - you're at risk of data changing under your feet here.

Yeah sorry missed that, will change it. :)
Thanks for taking time to review! :)

Reply:
Eurgh. I hadn't realised we weren't enforcing non-cyclicity - which I guess is hard without always forcing the transitive closure on dependency additional - and is itself prone to exponential blow up.

Ok. Let's think about this slightly differently - does it ever make sense to have an open pull request from a deleted branch? Probably not, even if that request has dependencies that it should prevent its closure - the fact that the code isn't there means it must be closed despite this.

So, if it doesn't break the database and the app layer to force it closed we should force it closed in spite of its dependencies. (I'm not sure for definite if this is the case - but I'd be surprised if it was the case.) 

I guess therefore either issue.ChangeStatus should have additional optional argument to force or, there needs to be a ForceClose function. I could imagine a possible use for a ForceClose for cases where there has been a rogue user - although it might make sense in such situations to just delete everything they've done.

(I appreciate that I'm nitpicking over what will be usually one PR but people have a tendency to deliberately break stuff.)